Sort & optimize d3ad0ne rules

0xVavaldi opened the pull request:
Recreated a new PR based on  https://github.com/hashcat/hashcat/pull/3795 reported issues and not wanting to dirty up the commits on rules.
Looks like I made the mistake of assuming a specific behavior when it came to the O rule.
Hashcat does not delete if the specified target goes out of bounds. My code deleted until the 'end'

Please verify & validate. This should be a lot easier to go through

Example: echo password | hashcat --stdout -j "O65" --force
Output: password
RuleProcessorY output: passwo

I've adjusted the behavior. Additionally, I added a feature to keep the same order (which I mixed up due to threading in the previous commit). Ignoring all previous changes these should be the new suggestions:

Before: 34156
After: 34109
same-op Optimized: 184
similar-op Removed: 47

```
Before:	$*	@*	+3	
After:	@*	+3	
Before:	$`	R5	o5Z	
After:	$`	o5Z	
Before:	$z	o00	T0	
After:	$z	o00	
Before:	'5	o2W	o21	
After:	'5	o21	
Before:	*36	+4	o4z	
After:	*36	o4z	
Before:	*5A	O08	D2	
After:	O08	D2	
Before:	*70	O05	o28	
After:	O05	o28	
Before:	*71	C	l	
After:	*71	l	
Before:	*7A	D4	'5	
After:	D4	'5	
Before:	*89	'5	$g	
After:	'5	$g	
Before:	*A0	-7	'5	
After:	*A0	'5	
Before:	*9A	$m	O83	
After:	$m	O83	
Before:	*AB	T7	'A	
After:	T7	'A	
Before:	+4	'4	l	
After:	'4	l	
Before:	+4	*24	O03	
After:	*24	O03	
Before:	+5	+0	D5	
After:	+0	D5	
Before:	+5	$x	O06	
After:	$x	O06	
Before:	+6	O06	o03	
After:	O06	o03	
Before:	+6	^*	'5	
After:	^*	'5	
Before:	+A	R5	'9	
After:	R5	'9	
Before:	-1	$R	@R	
After:	-1	@R	
Before:	-0	o0f	q	
After:	o0f	q	
Before:	-3	D3	i0F	
After:	D3	i0F	
Before:	-2	sw;	D2	
After:	sw;	D2	
Before:	-3	O86	'8	
After:	-3	'8	
Before:	-3	T3	u	
After:	-3	u	
Before:	-9	-7	'8	
After:	-7	'8	
Before:	-8	'8	$I	
After:	'8	$I	
Before:	-7	o77	i5I	
After:	o77	i5I	
Before:	@0	C	C	
After:	@0	C	
Before:	@2	T7	u	
After:	@2	u	
Before:	@e	i6g	'5	
After:	@e	'5	
Before:	@b	c	l	
After:	@b	l	
Before:	@i	t	u	
After:	@i	u	
Before:	C	'5	u	
After:	'5	u	
Before:	C	C	i6Z	
After:	C	i6Z	
Before:	C	o3G	C	
After:	C	o3G	
Before:	C	u	D9	
After:	u	D9	
Before:	D1	f	r	
After:	D1	f	
Before:	D6	'6	^y	
After:	'6	^y	
Before:	L0	$O	[	
After:	$O	[	
Before:	DB	C	'B	
After:	C	'B	
Before:	L0	+3	o0$	
After:	+3	o0$	
Before:	L0	[	o78	
After:	[	o78	
Before:	L1	o1t	i4b	
After:	o1t	i4b	
Before:	L4	+5	o49	
After:	+5	o49	
Before:	L8	$U	O45	
After:	$U	O45	
Before:	R0	[	o58	
After:	[	o58	
Before:	LA	*B5	'8	
After:	*B5	'8	
Before:	R2	D2	^b	
After:	D2	^b	
Before:	R2	o2n	^c	
After:	o2n	^c	
Before:	R8	$x	O72	
After:	$x	O72	
Before:	R7	+8	o7V	
After:	+8	o7V	
Before:	T1	*34	o1_	
After:	*34	o1_	
Before:	T1	O89	c	
After:	O89	c	
Before:	T2	C	D6	
After:	C	D6	
Before:	T1	k	C	
After:	k	C	
Before:	T0	u	o6k	
After:	u	o6k	
Before:	T3	o3i	u	
After:	o3i	u	
Before:	T2	o94	u	
After:	o94	u	
Before:	T3	c	i4G	
After:	c	i4G	
Before:	T4	O56	u	
After:	O56	u	
Before:	T6	*74	c	
After:	*74	c	
Before:	T9	o96	[	
After:	o96	[	
Before:	T7	o5x	u	
After:	o5x	u	
Before:	T5	O25	u	
After:	O25	u	
Before:	T7	c	O39	
After:	c	O39	
Before:	T8	o11	l	
After:	o11	l	
Before:	Z1	o1L	o1c	
After:	Z1	o1c	
Before:	Z3	TA	'9	
After:	Z3	'9	
Before:	Z5	K	i3f	
After:	Z5	i3f	
Before:	Z5	K	*17	
After:	Z5	*17	
Before:	^\	R2	o2j	
After:	^\	o2j	
Before:	^\	z1	T1	
After:	^\	z1	
Before:	^f	z4	k	
After:	^f	z4	
Before:	c	l	L4	
After:	l	L4	
Before:	c	so&	c	
After:	c	so&	
Before:	c	u	Z4	
After:	u	Z4	
Before:	d	o3b	o3b	
After:	d	o3b	
Before:	f	r	D8	
After:	f	D8	
Before:	f	r	i4q	
After:	f	i4q	
Before:	f	r	+5	
After:	f	+5	
Before:	i0<	Z2	K	
After:	i0<	Z2	
Before:	i1*	L9	'6	
After:	i1*	'6	
Before:	i33	^H	'4	
After:	^H	'4	
Before:	i4C	*87	'6	
After:	i4C	'6	
Before:	i4u	+0	o0r	
After:	i4u	o0r	
Before:	i5_	OB3	'9	
After:	i5_	'9	
Before:	i5q	'4	-3	
After:	'4	-3	
Before:	i6[	-2	'5	
After:	-2	'5	
Before:	i7d	'6	@2	
After:	'6	@2	
Before:	i9B	l	l	
After:	i9B	l	
Before:	l	o3p	u	
After:	o3p	u	
Before:	o0!	o2^	o25	
After:	o0!	o25	
Before:	l	u	OB2	
After:	u	OB2	
Before:	l	u	O03	
After:	u	O03	
Before:	l	u	i5C	
After:	u	i5C	
Before:	o0a	[	i2b	
After:	[	i2b	
Before:	o0j	stg	[	
After:	stg	[	
Before:	o0y	$C	D0	
After:	$C	D0	
Before:	o11	D1	^O	
After:	D1	^O	
Before:	o1T	t	C	
After:	o1T	C	
Before:	o1u	$y	o1u	
After:	$y	o1u	
Before:	o29	o65	T2	
After:	o29	o65	
Before:	o2'	^*	T3	
After:	o2'	^*	
Before:	o21	D2	snq	
After:	D2	snq	
Before:	o2 	u	o2a	
After:	u	o2a	
Before:	o2e	o1]	o1&	
After:	o2e	o1&	
Before:	o2k	o2i	[	
After:	o2i	[	
Before:	o31	T3	O61	
After:	o31	O61	
Before:	o39	C	o3!	
After:	C	o3!	
Before:	o3T	o3n	*35	
After:	o3n	*35	
Before:	o3g	*42	o3b	
After:	*42	o3b	
Before:	o4'	$e	D4	
After:	$e	D4	
Before:	o4#	'4	$3	
After:	'4	$3	
Before:	o4\	o2[	T2	
After:	o4\	o2[	
Before:	o46	o0Y	D4	
After:	o0Y	D4	
Before:	o4f	o4J	*20	
After:	o4J	*20	
Before:	o4[	o4m	r	
After:	o4m	r	
Before:	o50	o5!	t	
After:	o5!	t	
Before:	o6o	O13	o3v	
After:	O13	o3v	
Before:	o6j	'5	$o	
After:	'5	$o	
Before:	o6s	i6#	T6	
After:	o6s	i6#	
Before:	o7K	$b	O53	
After:	$b	O53	
Before:	o7b	$j	O26	
After:	$j	O26	
Before:	o8L	^o	'6	
After:	^o	'6	
Before:	o8X	o8C	O61	
After:	o8C	O61	
Before:	o8{	D5	o7a	
After:	D5	o7a	
Before:	o8|	'7	-6	
After:	'7	-6	
Before:	oB2	i6h	'9	
After:	i6h	'9	
Before:	oB/	R5	'8	
After:	R5	'8	
Before:	p2	*A5	'4	
After:	p2	'4	
Before:	p3	oAo	'7	
After:	p3	'7	
Before:	q	*8B	'5	
After:	q	'5	
Before:	r	^7	T0	
After:	r	^7	
Before:	s3W	+5	D5	
After:	s3W	D5	
Before:	sdf	T5	u	
After:	sdf	u	
Before:	sh2	l	c	
After:	sh2	c	
Before:	sm[	C	c	
After:	sm[	c	
Before:	spG	l	C	
After:	spG	C	
Before:	spY	t	l	
After:	spY	l	
Before:	t	o1O	u	
After:	o1O	u	
Before:	t	T1	o1h	
After:	t	o1h	
Before:	t	c	^Z	
After:	c	^Z	
Before:	t	i1a	u	
After:	i1a	u	
Before:	t	l	+A	
After:	l	+A	
Before:	t	o1C	c	
After:	o1C	c	
Before:	t	s6z	l	
After:	s6z	l	
Before:	t	u	i7Z	
After:	u	i7Z	
Before:	u	^u	c	
After:	^u	c	
Before:	u	l	s0i	
After:	l	s0i	
Before:	u	c	o3Q	
After:	c	o3Q	
Before:	u	c	@g	
After:	c	@g	
Before:	u	o4p	c	
After:	o4p	c	
Before:	u	l	s0?	
After:	l	s0?	
Before:	u	i88	l	
After:	i88	l	
Before:	O21	R1	o1x	
After:	O21	o1x	
Before:	O23	-2	o28	
After:	O23	o28	
Before:	O25	RA	'7	
After:	O25	'7	
Before:	O31	i73	'5	
After:	O31	'5	
Before:	O25	T6	'5	
After:	O25	'5	
Before:	O34	C	u	
After:	O34	u	
Before:	O45	oAX	'9	
After:	O45	'9	
Before:	O61	i3p	'7	
After:	i3p	'7	
Before:	O63	i82	T8	
After:	O63	i82	
Before:	O72	T7	o74	
After:	O72	o74	
Before:	O89	+9	'9	
After:	O89	'9	
Before:	O87	'8	o6u	
After:	'8	o6u	
Before:	OB1	D1	'8	
After:	D1	'8	
Before:	z1	*10	-B	
After:	z1	-B	
Before:	z1	$?	k	
After:	z1	$?	
Before:	O94	T9	o9g	
After:	O94	o9g	
Before:	z1	R7	k	
After:	z1	R7	
Before:	z1	k	^H	
After:	z1	^H	
Before:	z2	o6n	k	
After:	z2	o6n	
Before:	z2	i6g	k	
After:	z2	i6g	
Before:	z3	$A	k	
After:	z3	$A	
Before:	z3	k	sgc	
After:	z3	sgc	
Before:	z3	k	'7	
After:	z3	'7	
Before:	z3	ssz	*23	
After:	z3	ssz	
Before:	z4	R8	*21	
After:	z4	R8	
Before:	z5	k	i6d	
After:	z5	i6d	
Before:	z5	+8	o8f	
After:	z5	o8f	
Before:	z5	o6d	*15	
After:	z5	o6d	
Kept:	$+	[
Deleted:	D0	$+
Kept:	$*	D0
Deleted:	[	$*
Kept:	$0	O21
Deleted:	D2	$0
Kept:	$1	[
Deleted:	[	$1
Kept:	$1	O51
Deleted:	D5	$1
Kept:	$9	[
Deleted:	[	$9
Kept:	$?	[
Deleted:	[	$?
Kept:	$A	[
Deleted:	[	$A
Kept:	$X	[
Deleted:	[	$X
Kept:	$a	O31
Deleted:	D3	$a
Kept:	$b	[
Deleted:	[	$b
Kept:	$a	O91
Deleted:	D9	$a
Kept:	$k	[
Deleted:	[	$k
Kept:	*98	'9
Deleted:	O81	'9
Kept:	-6
Deleted:	-6
Kept:	-8
Deleted:	-8
Kept:	@7	[
Deleted:	D0	@7
Kept:	@9	[
Deleted:	[	@9
Kept:	@3	z2
Deleted:	z2	@3
Kept:	@7	z1
Deleted:	z1	@7
Kept:	@f	z1
Deleted:	z1	@f
Kept:	@h	[
Deleted:	[	@h
Kept:	@z	z1
Deleted:	z1	@z
Kept:	@x	[
Deleted:	D0	@x
Kept:	C	Z4
Deleted:	Z2	C	Z2
Kept:	C	Z2
Deleted:	Z2	C
Kept:	o0V
Deleted:	D0	^V
Kept:	o0u
Deleted:	D0	^u
Kept:	C	Z5
Deleted:	Z5	C
Kept:	o0/
Deleted:	D0	^/
Kept:	o0G
Deleted:	D0	^G
Kept:	C	Z3
Deleted:	Z3	C
Kept:	D0	^N
Deleted:	i1N	[
Kept:	D0	}
Deleted:	}	D1
Kept:	Z1	[
Deleted:	[	Z1
Kept:	Z3	c
Deleted:	c	Z3
Kept:	o0&
Deleted:	[	^&
Kept:	o0L
Deleted:	[	^L
Kept:	o0-
Deleted:	[	^-
Kept:	o0g
Deleted:	[	^g
Kept:	o0O
Deleted:	[	^O
Kept:	o0P
Deleted:	[	^P
Kept:	o0a
Deleted:	[	^a
Kept:	o0p
Deleted:	[	^p
Kept:	o0=
Deleted:	[	^=
Kept:	o0%
Deleted:	[	^%
Kept:	O31	}
Deleted:	}	D4
```

0xVavaldi committed:
Sort & optimize d3ad0ne rules

jsteube commented:
I've picked some random modifications and they seem to work fine from function perspective.
There's a few things you might want to improve in your tool before we merge this.
1. Optimize for speed. Example:
Kept:	O31	}
Deleted:	}	D4
- To explain what I mean we can reduce to:
Kept:	O31
Deleted: D4
- But, keep the D4 instead of the O31, because D4 executes faster
For testing I usually use hash mode 900 in optimized mode with w3 and a single hash and rockyou.txt. Make sure to have enough rules in place, for instance:
$ perl -e 'print "D4\n" x 200000' > rule  
2. Either keep the original format or always add a space between function calls. This example is not wanted:
$i $n $g
$i$n$g
3. Do not insert tab characters. Use spaces it's more consistent across different editors and easier to paste

0xVavaldi committed:
Update d3ad0ne.rule

Optimized based on (rough) performance per rule function.

0xVavaldi committed:
Add spaces

Add spaces between function as described.

0xVavaldi commented:
```
Before:	$*	@*	+3	
After:	@*	+3	
Before:	$`	R5	o5Z	
After:	$`	o5Z	
Before:	$z	o00	T0	
After:	$z	o00	
Before:	'5	o2W	o21	
After:	'5	o21	
Before:	*36	+4	o4z	
After:	*36	o4z	
Before:	*5A	O08	D2	
After:	O08	D2	
Before:	*70	O05	o28	
After:	O05	o28	
Before:	*71	C	l	
After:	*71	l	
Before:	*7A	D4	'5	
After:	D4	'5	
Before:	*89	'5	$g	
After:	'5	$g	
Before:	*9A	$m	O83	
After:	$m	O83	
Before:	*A0	-7	'5	
After:	*A0	'5	
Before:	*AB	T7	'A	
After:	T7	'A	
Before:	+4	'4	l	
After:	'4	l	
Before:	+4	*24	O03	
After:	*24	O03	
Before:	+5	$x	O06	
After:	$x	O06	
Before:	+5	+0	D5	
After:	+0	D5	
Before:	+6	^*	'5	
After:	^*	'5	
Before:	+6	O06	o03	
After:	O06	o03	
Before:	+A	R5	'9	
After:	R5	'9	
Before:	-1	$R	@R	
After:	-1	@R	
Before:	-2	sw;	D2	
After:	sw;	D2	
Before:	-3	D3	i0F	
After:	D3	i0F	
Before:	-3	O86	'8	
After:	-3	'8	
Before:	-3	T3	u	
After:	-3	u	
Before:	-0	o0f	q	
After:	o0f	q	
Before:	-8	'8	$I	
After:	'8	$I	
Before:	-9	-7	'8	
After:	-7	'8	
Before:	-7	o77	i5I	
After:	o77	i5I	
Before:	@0	C	C	
After:	@0	C	
Before:	@2	T7	u	
After:	@2	u	
Before:	@e	i6g	'5	
After:	@e	'5	
Before:	@b	c	l	
After:	@b	l	
Before:	@i	t	u	
After:	@i	u	
Before:	C	'5	u	
After:	'5	u	
Before:	C	o3G	C	
After:	C	o3G	
Before:	C	C	i6Z	
After:	C	i6Z	
Before:	C	u	D9	
After:	u	D9	
Before:	D1	f	r	
After:	D1	f	
Before:	D6	'6	^y	
After:	'6	^y	
Before:	DB	C	'B	
After:	C	'B	
Before:	L0	$O	[	
After:	$O	[	
Before:	L0	+3	o0$	
After:	+3	o0$	
Before:	L0	[	o78	
After:	[	o78	
Before:	L1	o1t	i4b	
After:	o1t	i4b	
Before:	L4	+5	o49	
After:	+5	o49	
Before:	L8	$U	O45	
After:	$U	O45	
Before:	R0	[	o58	
After:	[	o58	
Before:	R2	D2	^b	
After:	D2	^b	
Before:	LA	*B5	'8	
After:	*B5	'8	
Before:	R2	o2n	^c	
After:	o2n	^c	
Before:	R7	+8	o7V	
After:	+8	o7V	
Before:	R8	$x	O72	
After:	$x	O72	
Before:	T1	*34	o1_	
After:	*34	o1_	
Before:	T1	k	C	
After:	k	C	
Before:	T2	C	D6	
After:	C	D6	
Before:	T0	u	o6k	
After:	u	o6k	
Before:	T2	o94	u	
After:	o94	u	
Before:	T1	O89	c	
After:	O89	c	
Before:	T3	c	i4G	
After:	c	i4G	
Before:	T4	O56	u	
After:	O56	u	
Before:	T3	o3i	u	
After:	o3i	u	
Before:	T9	o96	[	
After:	o96	[	
Before:	T5	O25	u	
After:	O25	u	
Before:	T6	*74	c	
After:	*74	c	
Before:	T7	o5x	u	
After:	o5x	u	
Before:	T8	o11	l	
After:	o11	l	
Before:	T7	c	O39	
After:	c	O39	
Before:	Z1	o1L	o1c	
After:	Z1	o1c	
Before:	Z3	TA	'9	
After:	Z3	'9	
Before:	Z5	K	i3f	
After:	Z5	i3f	
Before:	Z5	K	*17	
After:	Z5	*17	
Before:	^\	z1	T1	
After:	^\	z1	
Before:	^\	R2	o2j	
After:	^\	o2j	
Before:	^f	z4	k	
After:	^f	z4	
Before:	c	l	L4	
After:	l	L4	
Before:	c	so&	c	
After:	c	so&	
Before:	c	u	Z4	
After:	u	Z4	
Before:	d	o3b	o3b	
After:	d	o3b	
Before:	f	r	i4q	
After:	f	i4q	
Before:	f	r	D8	
After:	f	D8	
Before:	f	r	+5	
After:	f	+5	
Before:	i0<	Z2	K	
After:	i0<	Z2	
Before:	i1*	L9	'6	
After:	i1*	'6	
Before:	i33	^H	'4	
After:	^H	'4	
Before:	i4C	*87	'6	
After:	i4C	'6	
Before:	i4u	+0	o0r	
After:	i4u	o0r	
Before:	i5_	OB3	'9	
After:	i5_	'9	
Before:	i5q	'4	-3	
After:	'4	-3	
Before:	i6[	-2	'5	
After:	-2	'5	
Before:	i7d	'6	@2	
After:	'6	@2	
Before:	i9B	l	l	
After:	i9B	l	
Before:	l	o3p	u	
After:	o3p	u	
Before:	o0!	o2^	o25	
After:	o0!	o25	
Before:	l	u	i5C	
After:	u	i5C	
Before:	l	u	OB2	
After:	u	OB2	
Before:	l	u	O03	
After:	u	O03	
Before:	o0a	[	i2b	
After:	[	i2b	
Before:	o0j	stg	[	
After:	stg	[	
Before:	o0y	$C	D0	
After:	$C	D0	
Before:	o11	D1	^O	
After:	D1	^O	
Before:	o1T	t	C	
After:	o1T	C	
Before:	o1u	$y	o1u	
After:	$y	o1u	
Before:	o2 	u	o2a	
After:	u	o2a	
Before:	o2'	^*	T3	
After:	o2'	^*	
Before:	o29	o65	T2	
After:	o29	o65	
Before:	o21	D2	snq	
After:	D2	snq	
Before:	o2e	o1]	o1&	
After:	o2e	o1&	
Before:	o2k	o2i	[	
After:	o2i	[	
Before:	o31	T3	O61	
After:	o31	O61	
Before:	o3T	o3n	*35	
After:	o3n	*35	
Before:	o3g	*42	o3b	
After:	*42	o3b	
Before:	o39	C	o3!	
After:	C	o3!	
Before:	o4'	$e	D4	
After:	$e	D4	
Before:	o46	o0Y	D4	
After:	o0Y	D4	
Before:	o4#	'4	$3	
After:	'4	$3	
Before:	o4\	o2[	T2	
After:	o4\	o2[	
Before:	o4[	o4m	r	
After:	o4m	r	
Before:	o4f	o4J	*20	
After:	o4J	*20	
Before:	o50	o5!	t	
After:	o5!	t	
Before:	o6o	O13	o3v	
After:	O13	o3v	
Before:	o6j	'5	$o	
After:	'5	$o	
Before:	o6s	i6#	T6	
After:	o6s	i6#	
Before:	o7K	$b	O53	
After:	$b	O53	
Before:	o7b	$j	O26	
After:	$j	O26	
Before:	o8L	^o	'6	
After:	^o	'6	
Before:	o8X	o8C	O61	
After:	o8C	O61	
Before:	o8|	'7	-6	
After:	'7	-6	
Before:	o8{	D5	o7a	
After:	D5	o7a	
Before:	oB2	i6h	'9	
After:	i6h	'9	
Before:	oB/	R5	'8	
After:	R5	'8	
Before:	p2	*A5	'4	
After:	p2	'4	
Before:	p3	oAo	'7	
After:	p3	'7	
Before:	q	*8B	'5	
After:	q	'5	
Before:	r	^7	T0	
After:	r	^7	
Before:	s3W	+5	D5	
After:	s3W	D5	
Before:	sdf	T5	u	
After:	sdf	u	
Before:	sh2	l	c	
After:	sh2	c	
Before:	sm[	C	c	
After:	sm[	c	
Before:	spG	l	C	
After:	spG	C	
Before:	spY	t	l	
After:	spY	l	
Before:	t	c	^Z	
After:	c	^Z	
Before:	t	i1a	u	
After:	i1a	u	
Before:	t	o1O	u	
After:	o1O	u	
Before:	t	T1	o1h	
After:	t	o1h	
Before:	t	o1C	c	
After:	o1C	c	
Before:	t	l	+A	
After:	l	+A	
Before:	t	s6z	l	
After:	s6z	l	
Before:	t	u	i7Z	
After:	u	i7Z	
Before:	u	c	@g	
After:	c	@g	
Before:	u	c	o3Q	
After:	c	o3Q	
Before:	u	^u	c	
After:	^u	c	
Before:	u	l	s0i	
After:	l	s0i	
Before:	u	i88	l	
After:	i88	l	
Before:	u	o4p	c	
After:	o4p	c	
Before:	u	l	s0?	
After:	l	s0?	
Before:	O21	R1	o1x	
After:	O21	o1x	
Before:	O23	-2	o28	
After:	O23	o28	
Before:	O25	RA	'7	
After:	O25	'7	
Before:	O25	T6	'5	
After:	O25	'5	
Before:	O31	i73	'5	
After:	O31	'5	
Before:	O34	C	u	
After:	O34	u	
Before:	O45	oAX	'9	
After:	O45	'9	
Before:	O61	i3p	'7	
After:	i3p	'7	
Before:	O63	i82	T8	
After:	O63	i82	
Before:	O72	T7	o74	
After:	O72	o74	
Before:	O89	+9	'9	
After:	O89	'9	
Before:	O87	'8	o6u	
After:	'8	o6u	
Before:	O94	T9	o9g	
After:	O94	o9g	
Before:	OB1	D1	'8	
After:	D1	'8	
Before:	z1	$?	k	
After:	z1	$?	
Before:	z1	R7	k	
After:	z1	R7	
Before:	z1	*10	-B	
After:	z1	-B	
Before:	z1	k	^H	
After:	z1	^H	
Before:	z2	o6n	k	
After:	z2	o6n	
Before:	z3	$A	k	
After:	z3	$A	
Before:	z2	i6g	k	
After:	z2	i6g	
Before:	z3	k	sgc	
After:	z3	sgc	
Before:	z3	k	'7	
After:	z3	'7	
Before:	z4	R8	*21	
After:	z4	R8	
Before:	z3	ssz	*23	
After:	z3	ssz	
Before:	z5	+8	o8f	
After:	z5	o8f	
Before:	z5	k	i6d	
After:	z5	i6d	
Before:	z5	o6d	*15	
After:	z5	o6d	
Kept:		[	$*
Deleted:	$*	D0
Kept:		$+	[
Deleted:	D0	$+
Kept:		$1	[
Deleted:	[	$1
Kept:		D2	$0
Deleted:	$0	O21
Kept:		D5	$1
Deleted:	$1	O51
Kept:		$9	[
Deleted:	[	$9
Kept:		$A	[
Deleted:	[	$A
Kept:		$?	[
Deleted:	[	$?
Kept:		$X	[
Deleted:	[	$X
Kept:		D9	$a
Deleted:	$a	O91
Kept:		$b	[
Deleted:	[	$b
Kept:		D3	$a
Deleted:	$a	O31
Kept:		$k	[
Deleted:	[	$k
Kept:		O81	'9
Deleted:	*98	'9
Kept:		-6
Deleted:	-6
Kept:		-8
Deleted:	-8
Kept:		@3	z2
Deleted:	z2	@3
Kept:		@7	[
Deleted:	D0	@7
Kept:		@7	z1
Deleted:	z1	@7
Kept:		@9	[
Deleted:	[	@9
Kept:		@h	[
Deleted:	[	@h
Kept:		@f	z1
Deleted:	z1	@f
Kept:		@x	[
Deleted:	D0	@x
Kept:		@z	z1
Deleted:	z1	@z
Kept:		C	Z2
Deleted:	Z2	C
Kept:		C	Z3
Deleted:	Z3	C
Kept:		C	Z5
Deleted:	Z5	C
Kept:		C	Z4
Deleted:	Z2	C	Z2
Kept:		o0G
Deleted:	D0	^G
Kept:		D0	^N
Deleted:	i1N	[
Kept:		o0/
Deleted:	D0	^/
Kept:		o0V
Deleted:	D0	^V
Kept:		o0u
Deleted:	D0	^u
Kept:		D0	}
Deleted:	}	D1
Kept:		Z1	[
Deleted:	[	Z1
Kept:		Z3	c
Deleted:	c	Z3
Kept:		o0-
Deleted:	[	^-
Kept:		o0p
Deleted:	[	^p
Kept:		o0&
Deleted:	[	^&
Kept:		o0g
Deleted:	[	^g
Kept:		o0=
Deleted:	[	^=
Kept:		o0O
Deleted:	[	^O
Kept:		o0L
Deleted:	[	^L
Kept:		o0P
Deleted:	[	^P
Kept:		o0a
Deleted:	[	^a
Kept:		o0%
Deleted:	[	^%
Kept:		}	D4
Deleted:	O31	}
```

Updated changes using the 'optimize for speed based on function'. I've taken the majority of rule functions into account for this. Feel free to point out notable inconsistencies.

 Additionally, I've added spaces to all functions for readability.
